src/lib.rs

```diff
@@ -415,83 +386,174 @@ fn finalize_fips_approval(
+                        Ok(t) => {
+                            let mut slot = config::Slot::new();
+                            slot.dbtype = Some(t.to_string());
+                            slot.dbpath = Some(filename);
+                            /* if this fails there will be no slots defined */
+                            let _ = gc.conf.add_slot(slot);
+                        }
+                        Err(_) => (),
+                    }
+                }
+                return gc;
+            }
         };
-    }
-    let filename = v[0].to_string();
 
-    let mut wstate = global_wlock!(noinitcheck STATE);
-    if !wstate.is_initialized() {
-        wstate.initialize();
+        GlobalConfig { conf: conf }
     }
 
-    /* if a slot num was not specified, pick the first free one,
-     * but ensure that the same file is not already open on an
-     * existing slot */
-    if v.len() == 1 {
-        let mut slot_high: i64 = -1;
-        let slots = wstate.get_slots_ids();
-        for s in &slots {
-            let sn = cast_or_ret!(i64 from *s);
-            if sn > slot_high {
-                slot_high = sn;
+    fn slot_from_init_args(&mut self, args: &str) -> Result<()> {
+        let mut slot = config::Slot::new();
+
+        let arguments: Vec<&str> = args.split(',').collect();
+        for arg in &arguments {
+            let comps: Vec<&str> = arg.splitn(2, '=').collect();
+            if comps.len() != 2 {
+                return Err(CKR_ARGUMENTS_BAD)?;
             }
-            match wstate.get_token_from_slot(*s) {
-                Ok(t) => {
-                    if filename.eq(t.get_filename()) {
-                        return CKR_CRYPTOKI_ALREADY_INITIALIZED;
+            match comps[0] {
+                "db" => {
+                    let dbv: Vec<&str> = comps[1].splitn(2, ':').collect();
+                    if dbv.len() == 1 {
+                        match storage::name_to_type(dbv[0]) {
+                            Ok(t) => {
+                                slot.dbtype = Some(t.to_string());
+                                slot.dbpath = Some(dbv[0].to_string());
+                            }
+                            Err(_) => {
+                                if dbv[0] == "memory" {
+                                    slot.dbtype =
+                                        Some(storage::MEMORYDB.to_string());
+                                } else {
+                                    return Err(CKR_ARGUMENTS_BAD)?;
+                                }
+                            }
+                        };
+                    } else {
+                        match dbv[0] {
+                            "sqlite" => {
+                                slot.dbtype =
+                                    Some(storage::SQLITEDB.to_string());
+                                slot.dbpath = Some(dbv[1].to_string());
+                            }
+                            "jsondb" => {
```

Comment:
should this be json or jsondb? As for now, all the other, sqlite and memory have the same name as their `storage::` counterparts, but json has different. Not sure if this is overlook, but having them in sync would make it easier to orient in that.
Some examples/documentation of the configuration file syntax, options and values would also be good to be able to understand how the conf files should look like and be constructed.

Reply:
I should change this to json indeed. I am preparing another PR based on this one that also brought this up to my attention.